### PR TITLE
Fix for point lights and translucency

### PR DESCRIPTION
**Pull request by e2002e**
There was a bug with translucent materials, when calling make_forward_base() and a point light is used, the number of arguments passed to sampleLight in the shader doesn't correspond with the definition.
Because the definition takes into account the SSRS and Microshadowing options, but not the part that I had to modify.
I'm not sure this is the proper way of setting the options, since it does only concern the point lights, I don't know about spots nor areas.
Seems effective though, with a glass candlestick and tesselation it looks like this:
None:
Capture d’écran du 2022-07-25 13-54-18
Microshadowing:
Capture d’écran du 2022-07-25 13-54-40
Microshadowing and ssrs:
Capture d’écran du 2022-07-25 13-55-02
ssrs without microshadowing:

Thing is microshadowing with ssr on seems to have no effect, we can see the wood being more lit hence some of it we can see through the glass ; but not on the glass itself.
Capture d’écran du 2022-07-25 13-58-58

### CODE REVIEW

File: blender/arm/material/make_mesh.py [hunk 1]
```diff
@@ -716,6 +730,13 @@ def make_forward_base(con_mesh, parse_opacity=False, transluc_pass=False):
             frag.write('  , true, spotData.x, spotData.y, spotDir, spotData.zw, spotRight')
         if '_VoxelShadow' in wrd.world_defs and '_VoxelAOvar' in wrd.world_defs:
             frag.write('  , voxels, voxpos')
+        if '_MicroShadowing' in wrd.world_defs:
+            frag.write(' , occspec.x')
+        if '_SSRS' in wrd.world_defs:
+       	    frag.add_include('std/ssrs.glsl')
```
Comment (MoritzBrueckner) [inline blender/arm/material/make_mesh.py:736]: What's the reason for including `ssrs.glsl` here? Also this line and line 739 have one space too much indentation (Python is sometimes very nitpicky about this).

File: blender/arm/material/make_mesh.py [hunk 2]
```diff
@@ -708,6 +707,21 @@ def make_forward_base(con_mesh, parse_opacity=False, transluc_pass=False):
...
+            frag.add_include('std/gbuffer.glsl')
+            frag.add_uniform('sampler2D gbuffer1')
+            frag.add_uniform('sampler2D gbufferD')
+            frag.add_uniform('vec3 eyeLook')
+            frag.add_uniform('vec2 cameraProj')
+            frag.add_in('vec2 texCoord')
+            frag.add_in('vec3 viewRay')
+            frag.write('vec4 g1 = textureLod(gbuffer1, texCoord, 0.0);')
+            frag.write('vec2 occspec = unpackFloat2(g1.a);')
+            frag.write('float depth = textureLod(gbufferD, texCoord, 0.0).r * 2.0 - 1.0;')
```
Comment (MoritzBrueckner) [inline blender/arm/material/make_mesh.py:720]: What are all these calculations for if you only use  `occspec` later?

**Review by MoritzBrueckner — CHANGES_REQUESTED** [review]
Nice!

Have you tested whether `gbufferD` actually contains valid values inside of the translucent shaders and not some garbage values or just (0,0,0)? I'm still not sure whether it does. It's still better this way than the current error message, but if you're unsure as well then maybe you want to add a little `# TODO` comment in the code so that someone debugging this code in the future knows to keep an eye on this?

File: blender/arm/material/make_mesh.py (hunk 2 above)
Comment (MoritzBrueckner) [reply blender/arm/material/make_mesh.py:720]: Seems like the GitHub previews is bit misleading, I meant `float depth`, `p`, `v` and everything above that is only required for those calculations.

**Comment by e2002e** [comment]
Yes, non needed computations, also right for the include.
It would be nice to set the sampler2D texture in the renderpath file but then I don't know how to set the shader as target.
rp_translucent uses 'translucent_resolve.frag' via Inc.hx

**Comment by e2002e** [comment]
Microshadowing works as expected:
without it and the lamp on the left
Capture d’écran du 2022-07-26 04-57-17
And activated:
Capture d’écran du 2022-07-26 04-57-40
What gets darker are the bottom parts so there is no aberration here.
The issue is that is does not seem to be of any effect when ssrs is turned on.

File: blender/arm/material/make_transluc.py [hunk 3]
```diff
@@ -18,7 +18,7 @@
 
 
 def make(context_id):
-	con_transluc = mat_state.data.add_context({ 'name': context_id, 'depth_write': False, 'compare_mode': 'less', 'cull_mode': 'clockwise', \
+	con_transluc = mat_state.data.add_context({ 'name': context_id, 'depth_write': False, 'compare_mode': 'less', 'cull_mode': 'none', \
```
Comment (MoritzBrueckner) [inline blender/arm/material/make_transluc.py:21]: What does this change fix?

Comment (e2002e) [reply blender/arm/material/make_transluc.py:21]: Nothing, I was trying to set up the glass material and forgot that.

**Review by MoritzBrueckner — APPROVED** [review]

**Comment by e2002e** [comment]
check the related issue sir.

**Comment by e2002e** [comment]
See, picture one is the candle's tray without microshadowing:

Capture d’écran du 2022-07-28 08-29-57

picture two shows lights from the flame, with microshadowing on:
Capture d’écran du 2022-07-28 08-29-44

**Comment by e2002e** [comment]
I made another change which is non-related and for which I had opened an issue.
Ocean needs to reflect something else than the sky, and this last commit fixes that.
It was a big pain because I though the code needed to be changed, but the issue was from the uninitialized buffers.
So that's a copy/paste, roughly put.
Capture d’écran du 2022-07-29 19-33-10

**Comment by QuantumCoderQC** [comment]
So does the ocean also reflect when SSR is disabled?

**Comment by e2002e** [comment]
> So does the ocean also reflect when SSR is disabled?

Yes. To not do so I guess we need to add an extra parameter to the water, maybe differentiate environment reflection from ssr.
I just pushed a commit addressing #2498
I don't know what the lines I've removed were for but that was the solution.

**Comment by MoritzBrueckner** [comment]
> I don't know what the lines I've removed were for but that was the solution.

Those lines ensure that the normals-related code is included in the shader before it's used by other parts of the shader. If I remember it correctly from my tests a while ago, the solution for this issue seemed to be a bit more complicated, but I currently have no time for it, maybe it is indeed that simple. If you haven't already, you should check the Git blame for changes around these lines and check whether your change reintroduces any of the old bugs that were fixed by code used in that function.

---

As I already mentioned in one of the related issues, please keep a single PR related to a single fix or feature, otherwise it complicates the repository's history and the discussion in the PR, makes it harder for Lubos to pick which changes to merge if he decides so, and all changes in this PR only show up as "Fix for point lights and translucency" in the release notes.

**Comment by e2002e** [comment]
Okay. I'm closing this one.

**Review by e2002e — COMMENTED** [review]
Issue appeared when an rgb mix's color is used for ArmoryPBR's normal input;
The normal is set with the texture before the later being declared.
I tried many modifications before noticing these two lines (I don't know what increasing a function does in python) and removing them fixes the issue.

**Review by e2002e — COMMENTED** [review]
The water pass is not working, I don't know how to reestablish it. I'm wiping it from the changes.

**Comment by MoritzBrueckner** [comment]
> I don't know what increasing a function does in python

`write_normal` is not a function, but a member variable of the `Shader` class ([source](https://github.com/armory3d/armory/blob/82ee3ae5efeabb0f508f646aa5a326e51f2d53ec/blender/arm/material/shader.py#L223)) which is kind of a stack-like counter in order to ensure that normals related code is added at the right position ([source](https://github.com/armory3d/armory/blob/82ee3ae5efeabb0f508f646aa5a326e51f2d53ec/blender/arm/material/shader.py#L350)). The counter is used instead of a simple boolean to make sure that leaving an "inner" normal parsing node doesn't reset the normal-parsing context of an outer node. So I guess that removing the counting and the positioning may break some other material setups in return...

**Comment by e2002e** [comment]
Ok, i'll check that once more. For now i've rolled back.
So the pull request is fixing voxel shadows and shadowmap atlas, and transculent related issues

**Comment by MoritzBrueckner** [comment]
> Something is puzzling me, the sampleLight() function is not used anywhere else than in make_clusters and make_mesh.

It's also used in the deferred light shaders (both standard render path and mobile), maybe that helps to resolve your confusion?

**Pull request closed without merging**